text/0000-bench-utils.md

```diff
@@ -0,0 +1,207 @@
+- Feature Name: black_box
+- Start Date: 2018-03-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC adds one function, `core::hint::black_box`, which is a hint to the
+optimizer to disable certain compiler optimizations. 
+
+# Motivation
+[motivation]: #motivation
+
+A tool for preventing compiler optimizations is widely useful. One application
+is writing synthetic benchmarks, where, due to the constrained nature of the
+benchmark, the compiler is able to perform optimizations that wouldn't otherwise
+trigger in practice. Another application is writing constant time code, where it
+is undesirable for the compiler to optimize certain operations depending on the
+context in which they are executed.
+
+The implementation of this function is backend-specific and currently requires
+inline assembly. No viable alternative is available in stable Rust.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+## `hint::black_box`
+
+The function:
+
+```rust
+/// An _unknown_ function that returns `x`.
+pub unsafe fn black_box<T>(x: T) -> T;
+```
+
+is an _unknown_ function, that is, a function that the compiler cannot make any
+assumptions about. It can potentially use `x` in any possible valid way that
+`unsafe` Rust code is allowed to, and requires the compiler to be maximally
+pessimistic in terms of optimizations. The compiler is still allowed to optimize
+the expression generating `x`. This function returns `x` and is a no-op in the
+virtual machine.
+
+For example ([`rust.godbolt.org`](https://godbolt.org/g/YP2GCJ)):
+
+```rust
+fn foo(x: i32) -> i32{ 
+  unsafe { hint::black_box(2 + x) };
+  3
+}
+let a = foo(2);
+```
+
+In the call to `foo(2)` the compiler is allowed to simplify the expression `2 + x` 
+down to `4`, but `4` must be stored into memory even though it is not read by
```

Comment:
"4 must be stored into memory" does not follow from the definition of `black_box` given here, it only has to be materialized in some way (e.g., in a register specified by the calling convention). Of course, the current (and likely most practical) implementation using inline assembly does force the value into memory, but that's neither here nor there.